klayout/drc/run_drc.py

```diff
@@ -0,0 +1,709 @@
+    ----------
+    drc_file : str
+        String that has the file full path to run.
+    path : str
+        String that holds the full path of the layout.
+    run_dir : str
+        String that holds the full path of the run location.
+    sws : dict
+        Dictionary that holds all switches that needs to be passed to the antenna checks.
+
+    Returns
+    -------
+    string
+        string that represent the path to the results output database for this run.
+
+    """
+
+    ## Using print because of the multiprocessing
+    logging.info(
+        "Running Global Foundries 180nm MCU {} checks on design {} on cell {}:".format(
+            path, drc_name, sws["topcell"]
+        )
+    )
+
+    layout_base_name = os.path.basename(path).split(".")[0]
+    new_sws = sws.copy()
+    report_path = os.path.join(
+        run_dir, "{}_{}.lyrdb".format(layout_base_name, drc_name)
+    )
+
+    new_sws["report"] = report_path
+    sws_str = build_switches_string(new_sws)
+    sws_str += f" -rd table_name={drc_name}"
+    # log_file = os.path.join(
+    #     run_dir, "{}_{}.log".format(layout_base_name, drc_name)
+    # )
+
+    run_str = f"klayout -b -r {drc_file} {sws_str}"
+
+    check_call(run_str, shell=True)
+
+    return report_path
+
+
+def run_parallel_run(
+    arguments: dict,
+    rule_deck_full_path: str,
+    layout_path: str,
+    switches: dict,
+    drc_run_dir: str,
+):
+    """
+    run_single_processor run the drc checks as in a multi-processing.
+
+    Parameters
+    ----------
+    arguments : dict
+        Dictionary that holds the arguments passed to the run_drc script.
+    rule_deck_full_path : str
+        String that holds the path of the rule deck files.
+    layout_path : str
+        Path to the target layout.
+    switches : dict
+        Dictionary that holds all the switches that will be passed to klayout run.
+    drc_run_dir : str
+        Path to the run location.
+    """
+
+    list_rule_deck_files = dict()
+
+    ## Run Antenna if required.
+    if arguments["--antenna"]:
+        drc_path = os.path.join(rule_deck_full_path, "rule_decks", "antenna.drc")
+        list_rule_deck_files["antenna"] = drc_path
+
+    ## Run Density if required.
+    if arguments["--density"]:
+        drc_path = os.path.join(rule_deck_full_path, "rule_decks", "density.drc")
+        list_rule_deck_files["density"] = drc_path
+
+    ## list_res_db_files.append(run_check(drc_path, "antenna", layout_path, drc_run_dir, switches))
+    if not arguments["--table"]:
+        list_of_tables = get_list_of_tables(rule_deck_full_path)
+    else:
+        list_of_tables = arguments["--table"]
+
+    ## Generate run rule deck from template.
+    for t in list_of_tables:
+        drc_file = generate_drc_run_template(rule_deck_full_path, drc_run_dir, [t])
+        list_rule_deck_files[t] = drc_file
+
+    ## Run All DRC files.
+    list_res_db_files = []
+    with concurrent.futures.ThreadPoolExecutor(max_workers=os.cpu_count()) as executor:
```

Comment:
alternatively we could make those individual pytest tests and use https://pypi.org/project/pytest-xdist/ to run them concurrently.

Reply:
`run_drc.py` is not part of the test. It's a drc run. I want to avoid the use of `pytest`.

Reply:
I see I thought this was mainly here as an helper for the tests; do we expect user to call `run_drc` from the command line? wouldn't they usually simply run the deck thru klayout?

Reply:
@proppy There are 2 ways to run DRC that we support:
- Command line based using this script `run_drc.py`
- GUI based that we didn't add here so far. But plan to add that one here.

What you currently have here is only the first option.

Reply:
@proppy No need to have this for `run_drc.py`

Reply:
assuming the rules are in separate files, couldn't `run_drc` be simplified to run a single rule and rely on standard `Makefile` orchestration (`-j`) to enable running multiple check currently?

Reply:
Opened an issue here #12